src/app/shared/code-editor-list/code-editor-list.service.ts

```diff
@@ -0,0 +1,291 @@
+import { Injectable } from '@angular/core';
+import { DescriptorLanguageService } from '../entry/descriptor-language.service';
+import { SourceFile, ToolDescriptor, Validation } from '../swagger';
+import { FileCategory } from './code-editor-list.component';
+
+/**
+ * This service currently only handles CWL, WDL, NFL, and Galaxy. All plugin languages and SERVICE will not work correctly.
+ *
+ * @export
+ * @class CodeEditorListService
+ */
+@Injectable({
+  providedIn: 'root'
+})
+export class CodeEditorListService {
+  static readonly NEXTFLOW_CONFIG_PATH = '/nextflow.config';
+  static readonly NEXTFLOW_PATH = '/main.nf';
+  constructor() {}
+  /**
+   * Determines whether to show the current sourcefile based on the descriptor type and tab
+   *
+   * @static
+   * @param {(SourceFile.TypeEnum | null | undefined)} sourcefileType
+   * @param {FileCategory} fileCategory
+   * @param {ToolDescriptor.TypeEnum} descriptorType
+   * @returns {boolean} Whether to show sourcefile or not
+   * @memberof CodeEditorListService
+   */
+  static showSourcefile(
+    sourcefileType: SourceFile.TypeEnum | null | undefined,
+    fileCategory: FileCategory | null | undefined,
+    descriptorType: ToolDescriptor.TypeEnum | null | undefined
+  ): boolean {
+    if (!sourcefileType || !fileCategory || !descriptorType) {
+      return true;
+    }
+    switch (fileCategory) {
+      case 'dockerfile': {
+        return true;
+      }
+      case 'descriptor': {
+        return (
+          (descriptorType === ToolDescriptor.TypeEnum.CWL && sourcefileType === SourceFile.TypeEnum.DOCKSTORECWL) ||
+          (descriptorType === ToolDescriptor.TypeEnum.WDL && sourcefileType === SourceFile.TypeEnum.DOCKSTOREWDL) ||
+          (descriptorType === ToolDescriptor.TypeEnum.NFL &&
+            (sourcefileType === SourceFile.TypeEnum.NEXTFLOW || sourcefileType === SourceFile.TypeEnum.NEXTFLOWCONFIG)) ||
+          (descriptorType === ToolDescriptor.TypeEnum.GXFORMAT2 && sourcefileType === Validation.TypeEnum.DOCKSTOREGXFORMAT2)
+        );
+      }
+      case 'testParam': {
+        return (
+          (descriptorType === ToolDescriptor.TypeEnum.CWL && sourcefileType === SourceFile.TypeEnum.CWLTESTJSON) ||
+          (descriptorType === ToolDescriptor.TypeEnum.WDL && sourcefileType === SourceFile.TypeEnum.WDLTESTJSON) ||
+          (descriptorType === ToolDescriptor.TypeEnum.NFL && sourcefileType === SourceFile.TypeEnum.NEXTFLOWTESTPARAMS) ||
+          (descriptorType === ToolDescriptor.TypeEnum.GXFORMAT2 && sourcefileType === Validation.TypeEnum.GXFORMAT2TESTFILE)
+        );
+      }
+    }
+  }
+
+  static isPrimaryDescriptor(path: string | null): boolean {
+    if (!path) {
+      return false;
+    }
+    const primaryDescriptors = [
+      '/Dockstore.cwl',
+      '/Dockstore.wdl',
+      CodeEditorListService.NEXTFLOW_CONFIG_PATH,
+      CodeEditorListService.NEXTFLOW_PATH
```

Comment:
What about Galaxy?

Reply:
Doesn't really seem to affect it for some reason.

Reply:
or maybe it does...hmm